Align macOS and iPadOS fullscreen keyboard behavior

stwrt opened the pull request:
#### 2ae1d5df810a01cd2246fc5c5bdd85184855ff6c
<pre>
Align macOS and iPadOS fullscreen keyboard behavior
<a href="https://bugs.webkit.org/show_bug.cgi?id=296391">https://bugs.webkit.org/show_bug.cgi?id=296391</a>
<a href="https://rdar.apple.com/problem/156522008">rdar://problem/156522008</a>

Reviewed by Brent Fulgham.

There has been a difference between keyboard behavior on macOS and iPadOS in fullscreen mode.
This patch brings these two back into alignment.

* Source/WebCore/PAL/pal/system/ios/Device.cpp
* Source/WebCore/PAL/pal/system/ios/Device.h
* Source/WebCore/PAL/pal/system/ios/UserInterfaceIdiom.h
* Source/WebCore/PAL/pal/system/ios/UserInterfaceIdiom.mm
* Source/WebKit/Shared/UserInterfaceIdiom.serialization.in
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp

Canonical link: <a href="https://commits.webkit.org/297962@main">https://commits.webkit.org/297962@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/07d33b34cc4c5ff6b7ad8db26f869df3e34eff61

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113597 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33304 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23744 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119768 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64364 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115486 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33911 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41874 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86399 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41463 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116545 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27067 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102095 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66739 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26320 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20216 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63494 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96448 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20301 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123003 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40600 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30315 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95251 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40991 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98306 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95004 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24246 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40137 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17938 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/36837 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40483 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45986 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40142 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43454 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41899 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -3879,28 +3879,6 @@ bool EventHandler::needsKeyboardEventDisambiguationQuirks() const
-bool EventHandler::isKeyEventAllowedInFullScreen(const PlatformKeyboardEvent& keyEvent) const
-{
-    RefPtr document = m_frame->document();
-    if (document->fullscreen().isFullscreenKeyboardInputAllowed())
```
`isFullscreenKeyboardInputAllowed()` is defined as `return fullscreenElement() && m_areKeysEnabledInFullscreen;`. It looks like `m_areKeysEnabledInFullscreen` gets set under certain circumstances where the ChromeClient wishes to reject keyboard input.

I wonder if we should revise this method to just be:
```
#if ENABLE(FULLSCREEN_API)
bool EventHandler::isKeyEventAllowedInFullScreen(const PlatformKeyboardEvent& keyEvent) const
{
    if (RefPtr document = m_frame->document()) {
        return document->fullscreen().isFullscreenKeyboardInputAllowed())

    return true;
}
```

brentfulgham replied:
@nt1m Can you remind us of why `DocumentFullscreen` has cases where `m_areKeysEnabledInFullscreen` might be set to false? I think this might just be set based on whether the Document fullscreen feature is enabled or not, so we probably don't want to lose this behavior ( unless that feature has been removed).

nt1m replied:
I'm pretty sure it's just iOS's spoofing restriction in general that disallows using the keyboard in fullscreen:
https://searchfox.org/wubkat/rev/7cd5229e664679bd5aa646fce5a6f6e9ae77aa20/Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp#176-186

@jernoble can say more, but it's to avoid webpages from being able to emulate the system UI in fullscreen and stealing data, so anything that remotely looks like typing will be disallowed.

brentfulgham replied:
@nt1m You can set that bit aside. We've evaluated that with the security team. I'm only interested in (1) web facing things that we need to make sure continue to work properly, and (2) WKWebView embed cases where we want this disabled.

The boolean set by the ChromeClient seems to be a case where client apps might need to be able to control (block) keyboard entry in fullscreen.

nt1m replied:
The only other check I see is `isDocumentFullscreenEnabled()` which we already check for in `DocumentFullscreen::requestFullscreen()` earlier on. If security is OK with iPhone fullscreen also allowing keystrokes, then Brandon's patch is fine (although there would be more to clean up). If iPhone should behave differently from iPad and preserve current behavior, then this patch should be revised

brentfulgham replied:
If it's possible to connect a keyboard to an iPhone, I don't see any reason we need to be more restrictive than iPad. I think we can treat them the same way. Surely someone will enjoy playing a keyboard-based game on their phone :-)

brentfulgham replied:
But I think you are wrong, @nt1m -- there's a bit more checking. It looks like in the old code we checked `DocumentFullScreen:: isFullscreenKeyboardInputAllowed()` which is controlled by various feature flag settings. I think we should retain this, in case a WebKit embedder wants to deny keyboard access for some reason (or else we should remove those feature flags!)

nt1m replied:
This is the only place that sets the flag you're mentioning: https://searchfox.org/wubkat/rev/7cd5229e664679bd5aa646fce5a6f6e9ae77aa20/Source/WebCore/dom/DocumentFullscreen.cpp#224-233
https://searchfox.org/wubkat/rev/7cd5229e664679bd5aa646fce5a6f6e9ae77aa20/Source/WebCore/dom/DocumentFullscreen.cpp#282

This is the check that those call sites refer to, which I do not see any API hooks for:
https://searchfox.org/wubkat/rev/7cd5229e664679bd5aa646fce5a6f6e9ae77aa20/Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp#176-186

brentfulgham commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -3964,9 +3942,6 @@ bool EventHandler::internalKeyEvent(const PlatformKeyboardEvent& initialKeyEvent
         }
-
-        if (!isKeyEventAllowedInFullScreen(initialKeyEvent))
-            return false;
```
I suggest you delete the implementation of `isKeyEventAllowedInFullScreen` above (like your patch currently does), but instead of deleting, just change this small check to be:

```
    if (!documentFullscreen->isFullscreenKeyboardInputAllowed())
        return false;
```

stwrt replied:
In the original code this is a true check, so even if we failed the below check

```
if (document->fullscreen().isFullscreenKeyboardInputAllowed())
   return true;
```

then either

```
 if (keyEvent.type() == PlatformKeyboardEvent::Type::Char) {
        if (keyEvent.text().length() != 1)
            return false;
        char16_t character = keyEvent.text()[0];
        return character == ' ';
    }
```

or

```
return (keyCode >= VK_BACK && keyCode <= VK_CAPITAL)
        || (keyCode >= VK_SPACE && keyCode <= VK_DELETE)
        || (keyCode >= VK_OEM_1 && keyCode <= VK_OEM_PLUS)
        || (keyCode >= VK_MULTIPLY && keyCode <= VK_OEM_8);
```

could still return true.

brentfulgham replied:
Oh, I see. Then I guess we should keep the patch in its current form, and perhaps make edits to DocumentFullscreen as @nt1m suggests may be needed.

It makes me wonder if we should be removing those feature flags, too, since we won't be using them for anything.

brentfulgham reviewed (changes requested):
I think this looks good, but I'd like to keep the check that makes sure the ChromeClient didn't turn keyboard input off (see my small adjustment).

nt1m reviewed (changes requested):
There's more cleanup to do in DocumentFullscreen depending on what the outcome of this patch should be

brentfulgham reviewed (approved):
Thank you for adjusting the approach here. I think this looks great. r=me.

stwrt committed:
Align macOS and iPadOS fullscreen keyboard behavior
https://bugs.webkit.org/show_bug.cgi?id=296391
rdar://problem/156522008

Reviewed by Brent Fulgham.

There has been a difference between keyboard behavior on macOS and iPadOS in fullscreen mode.
This patch brings these two back into alignment.

* Source/WebCore/PAL/pal/system/ios/Device.cpp
* Source/WebCore/PAL/pal/system/ios/Device.h
* Source/WebCore/PAL/pal/system/ios/UserInterfaceIdiom.h
* Source/WebCore/PAL/pal/system/ios/UserInterfaceIdiom.mm
* Source/WebKit/Shared/UserInterfaceIdiom.serialization.in
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp

Canonical link: https://commits.webkit.org/297962@main
